### DISCUSSION TITLE
How to efficiently retrieve objects with same meta/tag?

### DISCUSSION THREAD

**Discussion by dzhang28** [minio/minio#18432] [Q&A] [answered]
Hi guys,

I'm trying to find a way to retrieve objects with same meta/tag.
But from given API, it seems I can only achieve this goal by listing objects in all buckets and then check if meta/tag match from the returned list. ( Basically same as `mc find alias --tags=key=value` )
Then get all objects matched.

I believe this should be a very common use case.
Is there any better approach?
Thanks.

**Comment by klauspost** [comment, marked as answer]
There are no direct advantages, since a namespace scan is needed.

Use prefix-based separation, which is the only "indexed" operation.

**Reply by dzhang28** [reply]
Could you please elaborate on 'prefix-based separation' or is there any related document available?
Did you mean object name be prefixed with a specific tag?
Thanks

**Reply by klauspost** [reply]
https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html

**Comment by dzhang28** [comment]
I found some old posts recommending usage of ElasticSearch on meta data management.
I successfully setup ElasticSearch and enabled event notification.
But what annoys me is that user tag not contained in the notification message.

I used tcpflow to monitor http messages.

I added some tags to file '/test2/test.png'.
I can see my tags in message to MINIO as follows (tag : value, tag2 : value2, tag3 : v3):

```
127.000.000.001.56136-127.000.000.001.09000: PUT /test2/test.png?tagging=&versionId=null HTTP/1.1^M
Host: 127.0.0.1:9000^M
User-Agent: MinIO (linux; amd64) minio-go/v7.0.64 MinIO Console/(dev)^M
Content-Length: 171^M
Authorization: AWS4-HMAC-SHA256 Credential=TR67QGFKUXZPCO9SSQJE/20231115/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date;x-amz-security-token, Signature=bb029d8910b501e1d06e185c3bfafd42d381701442a4552f9c36598392479a26^M
Content-Md5: t6qUmAktg5E0IOMFJRt4Fg==^M
X-Amz-Content-Sha256: UNSIGNED-PAYLOAD^M
X-Amz-Date: 20231115T084407Z^M
X-Amz-Security-Token: eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJhY2Nlc3NLZXkiOiJUUjY3UUdGS1VYWlBDTzlTU1FKRSIsImV4cCI6MTcwMDA1NzkxMCwicGFyZW50IjoibWluaW9hZG1pbiJ9.W__IUYxUro1JrUNrMM69IIfIKdN8RyXhZRx67KJaWxHnybMCFfMCjqTppIZrGytqRAE2xFU69I1M5eB7NUwpHA^M
X-Forwarded-For: 112.46.135.130^M
^M
<Tagging><TagSet><Tag><Key>tag</Key><Value>value</Value></Tag><Tag><Key>tag2</Key><Value>value2</Value></Tag><Tag><Key>tag3</Key><Value>v3</Value></Tag></TagSet></Tagging>
```

But I cannot find those tags in event notification from MINIO to ElasticSearch:

```
127.000.000.001.60770-127.000.000.001.09200: PUT /bucketevents/_doc/lYcUPpCDavvW5G1mHM_XMSmICg3K6KvonM-Qx5RXaGw= HTTP/1.1^M
Host: localhost:9200^M
User-Agent: go-elasticsearch/7.17.10 (linux amd64; Go 1.21.3)^M
Content-Length: 1031^M
Authorization: Basic ZWxhc3RpYzptaW5pb2FkbWlu^M
Content-Type: application/json^M
X-Elastic-Client-Meta: es=7.17.10,go=1.21.3,t=7.17.10,hc=1.21.3^M
^M
{"Records":[{"eventVersion":"2.0","eventSource":"minio:s3","awsRegion":"","eventTime":"2023-11-15T08:44:07.634Z","eventName":"s3:ObjectCreated:PutTagging","userIdentity":{"principalId":"minioadmin"},"requestParameters":{"principalId":"minioadmin","region":"","sourceIPAddress":"112.46.135.130"},"responseElements":{"content-length":"0","x-amz-id-2":"dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8","x-amz-request-id":"1797BF6A49625CD6","x-minio-deployment-id":"c3b0b1fb-0648-4fc8-8aba-ae0e8bb0d9eb","x-minio-origin-endpoint":"http://192.168.0.131:9000"},"s3":{"s3SchemaVersion":"1.0","configurationId":"Config","bucket":{"name":"test2","ownerIdentity":{"principalId":"minioadmin"},"arn":"arn:aws:s3:::test2"},"object":{"key":"test.png","size":2334,"eTag":"661d4a8ee1cfba4a1c8736f5277d32fc","contentType":"image/png","userMetadata":{"content-type":"image/png"},"sequencer":"1797BE4EB215899E"}},"source":{"host":"112.46.135.130","port":"","userAgent":"MinIO (linux; amd64) minio-go/v7.0.64 MinIO Console/(dev)"}}]}
```


So is this intended behavior or a potential bug?

Thanks a lot.

My MINIO version is: 
minio version RELEASE.2023-11-01T18-37-25Z (commit-id=4b4a98d5e59354870325ad19703fba03d1b104c2)

**Reply by klauspost** [reply]
Generally we don't include request data in events.

Adding it is of course possible at the small expense of bloating events, but more seriously I could see it as a potentially unintended data leak if the tags contain sensitive data.
